### ISSUE TITLE
shellcheck crashes when locale mismatches with inputs

### ISSUE THREAD

**Issue by vapier** [koalaman/shellcheck#846]
not sure how best to handle this.  should fall back to binary mode or something ?  should at least give a nicer failure message :).

```$
$ echo '/touchégg' > test.sh
$ LC_ALL=en_US.UTF8 shellcheck -s sh ./test.sh
$ LC_ALL=C shellcheck -s sh ./test.sh
./test.sh: ./test.sh: hGetContents: invalid argument (invalid byte sequence)
```

**Comment by koalaman** [comment]
This should have been fixed since the last release (by forcing UTF-8 with a fallback on latin1). Can you try with the latest commit?

**Comment by kurahaupo** [comment]
> This should have been fixed since the last release (by forcing UTF-8 with a fallback on latin1). Can you try with the latest commit?

Both UTF-8 and Latin-1 have gaps, so that some bytes or byte-sequences are invalid. (Latin-1 does not use 128-159; UTF-8 requires each lead by to be followed by the correct number of follow bytes.)

Couldn't it just default to "raw octets", with all high-bit-set bytes counting as "not special" -- roughly equivalent to "alphabetic", which is the interpretation generally given to LANG=C.

**Comment by vapier** [comment]
i was testing with the latest release (0.4.5).  but as Martin said, neither UTF-8 nor Latin1 (ISO-8859-1) can represent all binary data, so forcing those alone wouldn't help.

another example are executable archives -- people (using projects like [`makeself`](http://www.megastep.org/makeself/)) can create shell scripts with tarballs appended to them so you can just run `./foo.run` to unpack+install things directly.  trying to run shellcheck on those in a UTF-8 environment fails:
```sh
$ LC_ALL=en_US.UTF8 shellcheck linuxq3apoint-1.32b-3.x86.run
linuxq3apoint-1.32b-3.x86.run: linuxq3apoint-1.32b-3.x86.run: hGetContents: invalid argument (invalid byte sequence)
```

**Comment by koalaman** [comment]
All byte sequences are valid Latin1. Whether the encoded code points are used for anything doesn't matter. 

ShellCheck also won't reject valid UTF-8 encoded code points that are currently undefined in Unicode.

**Comment by vapier** [comment]
sorry, you're correct of course.  the 0x80 to 0x9F range, while being unprintable, are valid codepoints -- they're the [Latin-1 Supplement](https://en.wikipedia.org/wiki/Latin-1_Supplement_(Unicode_block)) block.

so if shellcheck falls back to that on decode errors, it should be fine.

**Comment by koalaman** [comment]
Good call with the binary data! Turns out that ShellCheck failed for structurally valid UTF-8 in undefined planes (i.e. characters beyond 0x10FFFF).

This is fixed in 22c8625. ShellCheck will now happily chug through megabytes of binary garbage.
